src/expr.rs

```diff
@@ -645,14 +891,47 @@ fn rewrite_binary_op(context: &RewriteContext,
 
     // Second condition is needed in case of line break not caused by a
     // shortage of space, but by end-of-line comments, for example.
-    if rhs_result.len() > remaining_width || rhs_result.contains('\n') {
-        result.push('\n');
-        result.push_str(&make_indent(offset));
-    } else {
-        result.push(' ');
-    };
+    // Note that this is non-conservative, but its just to see if it's even
+    // worth trying to put everything on one line.
+    if rhs_result.len() + 2 + operator_str.len() < width && !rhs_result.contains('\n') {
+        // 1 = space between lhs expr and operator
+        if let Some(mut result) = lhs.rewrite(context,
+                                              width - 1 - operator_str.len(),
+                                              offset) {
+
+            result.push(' ');
+            result.push_str(&operator_str);
+            result.push(' ');
+
+            let remaining_width = width.checked_sub(last_line_width(&result)).unwrap_or(0);
 
+            if rhs_result.len() <= remaining_width {
+                result.push_str(&rhs_result);
+                return Some(result);
+            }
+
+            if let Some(rhs_result) = rhs.rewrite(context,
```

Comment:
I'm curious how this will play out vs just skipping to the case below. The tests actually pass without :-P

Let's review later.